src/Interpreters/Cache/FileCache.cpp

```diff
@@ -400,43 +480,137 @@ FileCache::getOrSet(
     size_t offset,
     size_t size,
     size_t file_size,
-    const CreateFileSegmentSettings & settings)
+    const CreateFileSegmentSettings & settings,
+    size_t file_segments_limit)
 {
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FilesystemCacheGetOrSetMicroseconds);
 
     assertInitialized();
 
-    const auto aligned_offset = roundDownToMultiple(offset, boundary_alignment);
-    const auto aligned_end = std::min(roundUpToMultiple(offset + size, boundary_alignment), file_size);
-    const auto aligned_size = aligned_end - aligned_offset;
+    FileSegment::Range range(offset, offset + size - 1);
 
-    FileSegment::Range range(aligned_offset, aligned_offset + aligned_size - 1);
+    const auto aligned_offset = roundDownToMultiple(range.left, boundary_alignment);
+    auto aligned_end_offset = std::min(roundUpToMultiple(offset + size, boundary_alignment), file_size) - 1;
 
-    auto locked_key = metadata.lockKeyMetadata(key, CacheMetadata::KeyNotFoundPolicy::CREATE_EMPTY);
+    chassert(aligned_offset <= range.left);
+    chassert(aligned_end_offset >= range.right);
 
+    auto locked_key = metadata.lockKeyMetadata(key, CacheMetadata::KeyNotFoundPolicy::CREATE_EMPTY);
     /// Get all segments which intersect with the given range.
-    auto file_segments = getImpl(*locked_key, range);
+    auto file_segments = getImpl(*locked_key, range, file_segments_limit);
+
+    if (file_segments_limit)
+    {
+        chassert(file_segments.size() <= file_segments_limit);
+        if (file_segments.size() == file_segments_limit)
+            range.right = aligned_end_offset = file_segments.back()->range().right;
```

Comment:
No, here the case means that in `[aligned_left_offset, range.right]` there is already file_segments_limit file segments found in cache, so there is no point checking the right part till `aligned_end_offset`, because it goes beyond limit. We do not loose alignment guarantee here, because `file_segments` here is what is already present in cache and it is enough for us for know. Then in CachedReadBuffer we will see that our current batch is over - we will return here in `getOrSet` with `range.right` as new left offset (it will now be `range.left`) and will get all the part till that `aligned_end_offset` in this second batch (or some `n'th` batch). E.g. we just delay taking this aligned right part until the next batch of file segments. This delay only takes place when `getImpl` (e.g. file segments which already in cache and intersect with requested range) already has more than `file_segments_limit` file segments.

Reply:
but if `offset + size == read_until_position` we won't return afaiu

Reply:
> but if offset + size == read_until_position we won't return afaiu

No. If we return on that case and do not request aligned part - it means that even if `file_segments_limit = 0`, e.g. unlimited, then we would not have used this aligned left last as well, which was the same before this PR.
The aligned right part makes sense and is actually being used only if we have uncovered right offset, e.g.
```
[_____________][                                        ]
                          ^                                         ^
                          range.end                          aligned_end
___]    [__]                         <--- cache state
                  [____]
                    ^ uncovered right part
```
as in this case only we create a bigger aligned part containing uncovered part + aligned part. Then the case `offset + size == read_until_position` does not matter, because if we read until `range.right`, we will already request aligned part. The exception is - if uncovered part takes exactly n * max_file_segment_size bytes and n == file_segments_limit. In this case we will indeed return without taking aligned part, but it was always the same, because there was a while loop: https://github.com/ClickHouse/ClickHouse/blob/b9cc1580e0357a0ca2fa761381797c3737677f55/src/Interpreters/Cache/FileCache.cpp#L432-L433
which was correct.

Reply:
I didn't realise these nuances before. and now I think that maybe we indeed need to untangle reading from storage and reading by processors more.